rfc/0000-lock-file.md

```diff
@@ -0,0 +1,280 @@
+
+# Summary
+
+Provide a `Marker.lock` file that is used to pin the versions of the dependencies of the compiled lint crates.
+
+# Motivation
+
+Today `Cargo.lock` file for the internal cargo workspace created for building the lint crates isn't exposed to the users of Marker. This means that every time when `cargo marker` runs on a fresh CI environment the latest compatible versions of lint crates and their dependencies are built. This poses several problems outlined below. This RFC intends to propose a solution for them.
+
+## Compatibility fiasco
+
+The real world is cruel. People may release a breaking change by mistake. Even `cargo-semver-checks` can't catch all possible semver compatibility breakages. Moreover, it is enough to say that not everyone uses it and people often make mistakes.
+
+Some crate could also mistakenly or out-of-no-better-option rely on the behavior not included in its dependency semver guarantees. For example, some crates don't bump their major version if they increase the MSRV, or there are crates that simply don't care about the MSRV. There are so many reasons one could break the semver compatibility.
+
+## Security
+
+One may not want to always use the latest version of the dependencies. That increases the risk that any of the dependencies in the tree introduces a malicious build script in the new patch version that leaks CI credentials, for example.
+
+## Licensing
+
+There are a lot of different software licenses available in the wild. Their formulations can pose arbitrary requirements for the usage of the licensed code. Users should know if the dependencies of the lint crates that they compile pose any restrictions on their usage. If Marker always uses the latest compatible versions of the dependencies, then dependencies become highly dynamic. For example, a 3-rd party crate may add a new dependency with the license unacceptable for the end user. Without any lock file it is almost impossible to track which licenses appear in the dependency tree. Tools like `cargo deny` and `cargo audit` need a lock file to inspect the exact version of the crates in the dependency tree.
+
+Licensing terms can also change in the new versions of the dependencies. Moreover, they can change so drastically that the users may want to cease using the dependency. If Marker always uses the latest compatible versions of the dependencies then it may start using the version of the dependency with unacceptable licensing terms.
+
+## Caching the cargo home on CI
+
+To save on the network bandwidth it's possible to run `cargo fetch` for the cargo workspace on one runner and share the fetched `$CARGO_HOME` with other runners via caching. If Marker's lint crates and their dependencies aren't in the `Cargo.lock` file, then `cargo fetch` won't fetch them making it not possible to run lint crate compilations with a `--frozen` flag on CI that prohibits networks access.
+
+# Guide-level explanation
+
+## What is a lint crate?
```

Comment:
Yeah, having the code organized similarly to this will be needed for the dispatch crate in the buildspace. I just bundled a bunch of my ideas for the "future Lint and LintPass API" look and tried to express them in this prototype of the user guide. Although, some impl considerations also leak in this text, but that's just for the RFC.

When thinking about having a shared lock file I decided to actually use the the crate that we have in the buildspace to combine all 3-rd party lint crates, and to make using `inventory` to register lints and lint passes would really help with that. This would remove a bunch of boilerplate and establish a clean-code pattern for organizing lint implementations into modules. That will need to be documented in a book and showcased in the lint crate template.

Reply:
Okay, I like the idea, so if you believe that this could work, we should give it a shot :D